### PR TITLE
Add adversarial negative option to cross-entropy loss for link prediction tasks

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#954 

*Description of changes:*
Add adversarial negative loss to cross-entropy loss for link prediction tasks.

Adversarial negative loss can boast the model performance. The following experiments follow the same setting as in https://github.com/classicsong/graphstorm/blob/focal-loss/examples/paper_experiments/training_configs/link_prediction/lp_also_buy_1_layer_rgcn_dot_entropy_loss_joint_32_negative.yaml.

With adversrial negative loss, the best test MRR increases from 0.380 to 0.8796 using the same setting as in [GraphStorm Paper](https://dl.acm.org/doi/pdf/10.1145/3637528.3671603).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/config/argument.py [hunk 1]
```diff
@@ -2523,6 +2523,17 @@ def lp_loss_func(self):
         # which means using the default evaluation metrics for different tasks.
         return BUILTIN_LP_LOSS_CROSS_ENTROPY
 
+    @property
+    def adversarial_temperature(self):
+        """ Temperature of link prediction adversarial cross entropy loss.
```
Comment (Oxfordblue7) [inline python/graphstorm/config/argument.py:2528]:
```suggestion
        """ Temperature of adversarial cross entropy loss for link prediction tasks.
```

File: python/graphstorm/config/argument.py [hunk 2]
```diff
@@ -3097,6 +3108,8 @@ def _add_link_prediction_args(parser):
             help="Link prediction loss function.")
     group.add_argument("--contrastive-loss-temperature", type=float, default=argparse.SUPPRESS,
             help="Temperature of link prediction contrastive loss.")
+    group.add_argument("--adversarial-temperature", type=float, default=argparse.SUPPRESS,
+            help="Temperature of link prediction adversarial cross entropy loss.")
```
Comment (Oxfordblue7) [inline python/graphstorm/config/argument.py:3112]:
```suggestion
            help="Temperature of adversarial cross entropy loss for link prediction tasks.")
```

File: docs/source/advanced/link-prediction.rst [hunk 3]
```diff
@@ -83,6 +85,28 @@ GraphStorm provides three options to compute training losses:
...
+    The loss of negative edges is as:
+
+    .. math::
+        \begin{eqnarray}
+            loss = - y \cdot \log score + (1 - y) \cdot \log (1 - score)
+            loss = softmax(score * adversarial_temperature) * loss
+        \end{eqnarray}
+
+    where ``y`` is 0, ``score`` is the score value of the negative edges computed by the score function and ``adversarial_temperature`` is a hyper-parameter.
+
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:109]: Add description for **Weighted Adversarial Cross Entropy Loss**?

File: docs/source/advanced/link-prediction.rst [hunk 4]
```diff
@@ -83,6 +85,28 @@ GraphStorm provides three options to compute training losses:
...
+    The loss of negative edges is as:
+
+    .. math::
+        \begin{eqnarray}
+            loss = - y \cdot \log score + (1 - y) \cdot \log (1 - score)
+            loss = softmax(score * adversarial_temperature) * loss
+        \end{eqnarray}
+
+    where ``y`` is 0, ``score`` is the score value of the negative edges computed by the score function and ``adversarial_temperature`` is a hyper-parameter.
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:108]: Add description about the final loss function, like ``loss = (avg(loss_pos) + avg(loss_neg)) / 2``.

File: docs/source/advanced/link-prediction.rst [hunk 5]
```diff
@@ -83,6 +85,28 @@ GraphStorm provides three options to compute training losses:
...
     where ``G`` is the training graph.
 
+
+* **Adversarial Cross Entropy Loss**: The adversarial cross entropy loss turns a link prediction task into a binary classification task. We treat positive edges as 1 and negative edges as 0. In addition, adversarial cross-entropy loss adjusts the loss value of each negative sample based on its degree of difficulty. This is enabled by setting the ``--adversarial_temperature`` config.
+
+    The loss of positive edges is as:
+
+    .. math::
+        \begin{eqnarray}
+            loss = - y \cdot \log score + (1 - y) \cdot \log (1 - score)
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:95]:
```suggestion
            loss_{pos} = - y \cdot \log score + (1 - y) \cdot \log (1 - score)
```

File: docs/source/advanced/link-prediction.rst [hunk 6]
```diff
@@ -83,6 +85,28 @@ GraphStorm provides three options to compute training losses:
...
+        \end{eqnarray}
+
+    where ``y`` is 1, ``score`` is the score value of the positive edges computed by the score function.
+
+    The loss of negative edges is as:
+
+    .. math::
+        \begin{eqnarray}
+            loss = - y \cdot \log score + (1 - y) \cdot \log (1 - score)
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:104]:
```suggestion
            loss_{neg} = - y \cdot \log score + (1 - y) \cdot \log (1 - score)
```

File: docs/source/advanced/link-prediction.rst [hunk 7]
```diff
@@ -83,6 +85,28 @@ GraphStorm provides three options to compute training losses:
...
+            loss = softmax(score * adversarial_temperature) * loss
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:105]:
```suggestion
            loss_{neg} = softmax(score * adversarial_temperature) * loss_{neg}
```

File: tests/unit-tests/test_config.py [hunk 8]
```diff
@@ -1034,6 +1034,19 @@ def create_lp_config(tmp_path, file_name):
     with open(os.path.join(tmp_path, file_name+"_fail_metric3.yaml"), "w") as f:
         yaml.dump(yaml_object, f)
 
+    yaml_object["gsf"]["link_prediction"] = {
+        "adversarial_temperature": 0.1, # eval metric must be string or list
```
Comment (Oxfordblue7) [inline tests/unit-tests/test_config.py:1038]:
```suggestion
        "adversarial_temperature": 0.1,
```

File: tests/unit-tests/test_config.py [hunk 9]
```diff
@@ -1034,6 +1034,19 @@ def create_lp_config(tmp_path, file_name):
...
+    }
+    with open(os.path.join(tmp_path, file_name+"_adv_temp.yaml"), "w") as f:
+        yaml.dump(yaml_object, f)
+
+    yaml_object["gsf"]["link_prediction"] = {
+        "lp_loss_func": BUILTIN_LP_LOSS_CONTRASTIVELOSS,
+        "adversarial_temperature": 0.1, # eval metric must be string or list
```
Comment (Oxfordblue7) [inline tests/unit-tests/test_config.py:1045]:
```suggestion
        "adversarial_temperature": 0.1,
```

File: docs/source/advanced/link-prediction.rst (hunk 3 above)
Comment (classicsong) [reply docs/source/advanced/link-prediction.rst:109]: added

File: docs/source/cli/model-training-inference/configuration-run.rst [hunk 10]
```diff
@@ -485,6 +485,12 @@ Link Prediction Task
     - Argument: ``--lp-loss-func contrastive``
     - Default value: ``cross_entropy``
 
+- **adversarial_temperature**: Enable adversarial cross entropy loss and set the ``adversarial_temperature`` hyper-parameter. Only work when ``lp_loss_func`` is set to ``cross_entropy``. More detials can be found on the :ref:`Link Prediction Loss Functions<_link-prediction-loss>`.
```
[the remark this answers is not in the document]
Comment (classicsong) [reply docs/source/cli/model-training-inference/configuration-run.rst:488]: https://graphstorm-xiang.readthedocs.io/en/latest/advanced/link-prediction.html#link-prediction-loss-functions

File: docs/source/advanced/link-prediction.rst [hunk 11]
```diff
@@ -49,43 +49,84 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+
+        loss_{pos} = - w * \log score
+
+    where ``score`` is the score value of the positive edges computed by the score function, ``w`` is the weight of each positive edge. The loss of the negative edges is the same as **Adversarial Cross Entropy Loss**.
+
+    The final loss is as:
+
+    .. math::
+
+        loss = (avg(loss_{pos}) + avg(loss_{neg})) / 2
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:124]:
```suggestion
        loss = \left(\mathrm{avg}(loss_{pos}) + \mathrm{avg}(loss_{neg})\right) / 2
```

File: docs/source/advanced/link-prediction.rst [hunk 12]
```diff
@@ -49,43 +49,84 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+
+        loss_{neg} = softmax(score * adversarial\_temperature) * loss_{neg}
+
+    where ``score`` is the score value of the negative edges computed by the score function and ``adversarial_temperature`` is a hyper-parameter.
+
+    The final loss is as:
+
+    .. math::
+
+        loss = (avg(loss_{pos}) + avg(loss_{neg})) / 2
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:110]:
```suggestion
        loss = \left(\mathrm{avg}(loss_{pos}) + \mathrm{avg}(loss_{neg})\right) / 2
```

File: docs/source/advanced/link-prediction.rst [hunk 13]
```diff
@@ -49,43 +49,84 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+
+        loss_{pos} = - \log score
+
+    where ``score`` is the score value of the positive edges computed by the score function.
+
+    The loss of negative edges is as:
+
+    .. math::
+
+        loss_{neg} = \log (1 - score)
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:100]:
```suggestion
```

File: docs/source/advanced/link-prediction.rst [hunk 14]
```diff
@@ -49,43 +49,84 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:101]:
```suggestion
```

File: docs/source/advanced/link-prediction.rst [hunk 15]
```diff
@@ -49,43 +49,84 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+    .. math::
+
+        loss = (avg(loss_{pos}) + avg(loss_{neg})) / 2
+
 * **Contrastive Loss**: The contrastive loss compels the representations of connected nodes to be similar while forcing the representations of disconnected nodes remains dissimilar. In the implementation, we use the score computed by the score function to represent the distance between nodes. When computing the loss, we group one positive edge with the ``N`` negative edges corresponding to it.The loss function is as follows:
 
     .. math::
+
         loss = -log(\dfrac{exp(pos\_score)}{\sum_{i=0}^N exp(score\_i)})
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:130]:
```suggestion
        loss = -\log \left( \dfrac{\exp(pos\_score)}{\sum_{i=0}^N \exp(score\_i)} \right)
```

File: docs/source/advanced/link-prediction.rst [hunk 16]
```diff
@@ -49,44 +49,85 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+    .. math::
+
+        loss = \left(\mathrm{avg}(loss_{pos}) + \mathrm{avg}(loss_{neg})\right) / 2
+
 * **Contrastive Loss**: The contrastive loss compels the representations of connected nodes to be similar while forcing the representations of disconnected nodes remains dissimilar. In the implementation, we use the score computed by the score function to represent the distance between nodes. When computing the loss, we group one positive edge with the ``N`` negative edges corresponding to it.The loss function is as follows:
 
     .. math::
-        loss = -log(\dfrac{exp(pos\_score)}{\sum_{i=0}^N exp(score\_i)})
+
+        loss = -log(\dfrac{exp(pos\_score)}{\sum_{i=0}^N \exp(score\_i)})
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:130]:
```suggestion
        loss = -\log \left( \dfrac{\exp(pos\_score)}{\sum_{i=0}^N \exp(score\_i)} \right)
```

File: docs/source/advanced/link-prediction.rst [hunk 17]
```diff
@@ -49,44 +49,85 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+
+        loss_{pos} = - w * \log score
+
+    where ``score`` is the score value of the positive edges computed by the score function, ``w`` is the weight of each positive edge. The loss of the negative edges is the same as **Adversarial Cross Entropy Loss**.
+
+    The final loss is as:
+
+    .. math::
+
+        loss = \left(\mathrm{avg}(loss_{pos}) + \mathrm{avg}(loss_{neg})\right) / 2
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:124]:
```suggestion
        loss = \dfrac{\mathrm{avg}(loss_{pos}) + \mathrm{avg}(loss_{neg})}{2}
```

File: docs/source/advanced/link-prediction.rst [hunk 18]
```diff
@@ -49,44 +49,85 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+
+        loss_{neg} &= \mathrm{softmax}(score * adversarial\_temperature) * loss_{neg}
+
+    where ``score`` is the score value of the negative edges computed by the score function and ``adversarial_temperature`` is a hyper-parameter.
+
+    The final loss is as:
+
+    .. math::
+
+        loss = \left(\mathrm{avg}(loss_{pos}) + \mathrm{avg}(loss_{neg})\right) / 2
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:110]:
```suggestion
        loss = \dfrac{\mathrm{avg}(loss_{pos}) + \mathrm{avg}(loss_{neg})}{2}
```

File: docs/source/advanced/link-prediction.rst [hunk 19]
```diff
@@ -49,44 +49,85 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+        loss_{pos} = - \log score
+
+    where ``score`` is the score value of the positive edges computed by the score function.
+
+    The loss of negative edges is as:
+
+    .. math::
+
+        loss_{neg} &= \log (1 - score)
+
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:101]:
```suggestion
```

File: docs/source/advanced/link-prediction.rst [hunk 20]
```diff
@@ -49,44 +49,85 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+        loss_{pos} = - \log score
+
+    where ``score`` is the score value of the positive edges computed by the score function.
+
+    The loss of negative edges is as:
+
+    .. math::
+
+        loss_{neg} &= \log (1 - score)
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:100]:
```suggestion
        loss_{neg} = \log (1 - score) \\
```

File: docs/source/advanced/link-prediction.rst [hunk 21]
```diff
@@ -49,44 +49,85 @@ GraphStorm provides two ways to compute link prediction scores: Dot Product and
...
+
+        loss_{neg} &= \mathrm{softmax}(score * adversarial\_temperature) * loss_{neg}
```
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:102]:
```suggestion
        loss_{neg} = \mathrm{softmax}(score * adversarial\_temperature) * loss_{neg}
```

File: docs/source/advanced/link-prediction.rst (hunk 20 above)
Comment (classicsong) [reply docs/source/advanced/link-prediction.rst:100]: `&` is used to align the equations.

File: docs/source/advanced/link-prediction.rst (hunk 19 above)
Comment (classicsong) [reply docs/source/advanced/link-prediction.rst:101]: Sphinx doc suggests to use empty line to separate two equations.

File: docs/source/advanced/link-prediction.rst (hunk 20 above)
Comment (Oxfordblue7) [reply docs/source/advanced/link-prediction.rst]: was trying center align

File: docs/source/advanced/link-prediction.rst
Comment (Oxfordblue7) [inline docs/source/advanced/link-prediction.rst:1]: Can we also use \mathrm for sum in line 38 and line 46?

**Review by Oxfordblue7 — APPROVED** [review]
